include/odp/api/spec/ipsec.h

```diff
@@ -1688,7 +1689,15 @@ int odp_ipsec_in(const odp_packet_t pkt_in[], int num_in,
  * Each input packet must have a valid value for these metadata (other metadata
  * is ignored):
  * - L3 offset: Offset to the first byte of the (outmost) IP header
- * - L4 offset: Offset to the L4 header if L4 checksum offload is requested
+ * - L3 type:   Type of the IP packet (ODP_PROTO_L3_TYPE_IPv4 or _IPv6).
+ *              Ignored if the tfc_dummy flag is set.
```

Comment:
There is no direct way of setting L3 type like there is for L3 offset (i.e. odp_packet_l3_offset_set()). So an application cannot set the L3 type for locally originated packets in any way other than by calling odp_packet_parse() which is heavy and seems to defeat the purpose of this change.

I am not sure if making L3 type settable and thus changing it from mere parse result to more general metadata would be good. One option would be to just drop the requirement of setting the L3 type.

Reply:
odp_packet_has_ipv6_set() or odp_packet_has_ipv4_set() is already available right ?

Reply:
Yes, as discussed, that could be used but it probably requires some clarification in the API spec.

Reply:
It would good to mention explicitly how L3 type is set (use odp_packet_has_ipv4_set()...)

Reply:
Maybe we should just drop the requirement for L3 type since the way to setting L3 type is non-obvious and it may actually require more than one call (e.g. to set IPv4 flag and to clear IPv6 flag). At least SW based implementations can just about as easily figure out L3 type from the IP header which is likely in the cache.

@nithind1988, is it ok for you if we remote the L3 type requirement from both functions? If yes, I will take it away from the next version and add your reviewed-by tag.

Reply:
Can we keep it given that for packets that are being forwarded, there is no need to set the L3 type explicitly again ?
I guess the concern is only when packet is being created newly by application.

And with respect meta data change, I guess the text can be like that is present for PKTIO Tx checksum generation.
and isn't it implicit that has IPv4 set => has_ipv6 is automatically cleared as both cannot coexist in single layer

`

> 
> * For correct operation, packet metadata must provide valid offsets and type       
> * flags for the appropriate layer 3 and layer 4 protocols.  L3 and L4 offsets      
> * can be updated with odp_packet_l3_offset_set() and odp_packet_l4_offset_set()    
> * calls. L3 and L4 type flags can be updated using odp_packet_has_*_set() calls    
> * For example, UDP checksum calculation needs both L3 and L4 types (IP and UDP) and
> * L3 and L4 offsets (to access IP and UDP headers), while IP checksum              
> * calculation only needs L3 type (IP) and L3 offset (to access IP header).         
> * When application (e.g. a switch) does not modify L3/L4 data and thus checksum    
> * does not need to be updated, checksum insertion should be disabled for optimal   
> * performance.

                                                                     
`

Reply:
> for packets that are being forwarded, there is no need to set the L3 type explicitly again ?

The fact that a packet is forwarded (and inbound packet parsing is enabled) is not sufficient. There must not be tunneling (e.g. IPv6 carried in any manner inside IPv4 or vice versa) or protocol conversion (e.g. from non-IP to IP) involved. And in some applications the majority of packets may be effectively locally originated.

> I guess the text can be like that is present for PKTIO Tx checksum generation.

I would either remove the L3 type requirement, or add this sentence: "L3 type can be updated using odp_packet_has_*_set() calls."

> isn't it implicit that has IPv4 set => has_ipv6 is automatically cleared as both cannot coexist in single layer

I do not think it is since linux-gen ODP and ODP-DPDK do not work that way. Maybe (?) the API could be changed so that setting one flag in a layer resets the other flags of the same layer.

But is there really some tangible benefit in requiring the L3 type? If not then maybe we do not need to add it now. I understood that you just need L4 offset for speeding up your implementation.

Reply:
>But is there really some tangible benefit in requiring the L3 type? If not then maybe we do not need to add it now. I understood that you just need L4 offset for speeding up your implementation.

Ok. I was taking w.r.t general semantics which could be useful for other implementations

>I would either remove the L3 type requirement, or add this sentence: "L3 type can be updated using odp_packet_has_*_set() calls."

Yes, l4 offset is sufficient for us for outbound inline for today. So we are fine either way.